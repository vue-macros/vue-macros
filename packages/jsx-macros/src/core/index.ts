import {
  babelParse,
  generateTransform,
  getLang,
  HELPER_PREFIX,
  MagicStringAST,
  walkAST,
  type CodeTransform,
} from '@vue-macros/common'
import type { OptionsResolved } from '..'
import { transformDefineComponent } from './define-component'
import { transformDefineExpose } from './define-expose'
import { transformDefineModel } from './define-model'
import { transformDefineSlots } from './define-slots'
import { transformDefineStyle } from './define-style'
import type {
  ArrowFunctionExpression,
  CallExpression,
  FunctionDeclaration,
  FunctionExpression,
  Node,
  Program,
} from '@babel/types'

export { restructure } from './restructure'

export type FunctionalNode =
  | FunctionDeclaration
  | FunctionExpression
  | ArrowFunctionExpression

export type DefineStyle = {
  expression: CallExpression
  isDeclaration: boolean
  lang: string
}

export type RootMapValue = {
  defineComponent?: CallExpression
  defineModel?: {
    expression: CallExpression
    isRequired: boolean
  }[]
  defineSlots?: CallExpression
  defineExpose?: CallExpression
  defineStyle?: DefineStyle[]
}

export function transformJsxMacros(
  code: string,
  id: string,
  importMap: Map<string, string>,
  options: OptionsResolved,
): CodeTransform | undefined {
  const s = new MagicStringAST(code)
  const ast = babelParse(s.original, getLang(id))
  const rootMap = getRootMap(ast, s, options)

  let defineStyleIndex = 0
  for (const [root, map] of rootMap) {
    map.defineStyle?.forEach((defineStyle) => {
      transformDefineStyle(defineStyle, defineStyleIndex++, root, s, importMap)
    })

    if (root === undefined) continue

    let propsName = `${HELPER_PREFIX}props`
    if (root.params[0]) {
      if (root.params[0].type === 'Identifier') {
        propsName = root.params[0].name
      } else if (root.params[0].type === 'ObjectPattern') {
        const lastProp = root.params[0].properties.at(-1)
        if (
          !map.defineComponent &&
          lastProp?.type === 'RestElement' &&
          lastProp.argument.type === 'Identifier'
        ) {
          propsName = lastProp.argument.name
        } else {
          s.appendRight(
            root.params[0].extra?.trailingComma
              ? (root.params[0].extra?.trailingComma as number) + 1
              : lastProp?.end || root.params[0].end! - 1,
            `${
              !root.params[0].extra?.trailingComma &&
              root.params[0].properties.length
                ? ','
                : ''
            } ...${HELPER_PREFIX}props`,
          )
        }
      }
    } else {
      s.appendRight(getParamsStart(root, s.original), propsName)
    }

    if (map.defineComponent) {
      transformDefineComponent(root, propsName, map, s, ast, options)
    }
    if (map.defineModel?.length) {
      map.defineModel.forEach(({ expression }) => {
        transformDefineModel(expression, propsName, s)
      })
    }
    if (map.defineSlots) {
      transformDefineSlots(map.defineSlots, propsName, s, options.lib)
    }
    if (map.defineExpose) {
      transformDefineExpose(
        map.defineExpose,
        propsName,
        root,
        s,
        options.lib,
        options.version,
      )
    }
  }

  return generateTransform(s, id)
}

function getRootMap(ast: Program, s: MagicStringAST, options: OptionsResolved) {
  const parents: (Node | undefined | null)[] = []
  const rootMap = new Map<FunctionalNode | undefined, RootMapValue>()
  walkAST<Node>(ast, {
    enter(node, parent) {
      parents.unshift(parent)
      const root = isFunctionalNode(parents[1]) ? parents[1] : undefined

      if (
        root &&
        parents[2]?.type === 'CallExpression' &&
        options.defineComponent.alias.includes(s.sliceNode(parents[2].callee))
      ) {
        if (!rootMap.has(root)) rootMap.set(root, {})
        if (!rootMap.get(root)!.defineComponent) {
          rootMap.get(root)!.defineComponent = parents[2]
        }
      }

      const expression =
        node.type === 'VariableDeclaration'
          ? node.declarations[0].init?.type === 'CallExpression' &&
            s.sliceNode(node.declarations[0].init.callee) === '$'
            ? node.declarations[0].init.arguments[0]
            : node.declarations[0].init
          : node.type === 'ExpressionStatement'
            ? node.expression
            : undefined
      if (!expression) return
      const macroExpression = getMacroExpression(expression, options)
      if (!macroExpression) return
      if (!rootMap.has(root)) rootMap.set(root, {})
      const macroName = s.sliceNode(
        macroExpression.callee.type === 'MemberExpression'
          ? macroExpression.callee.object
          : macroExpression.callee,
      )
      if (macroName) {
        if (options.defineModel.alias.includes(macroName)) {
          ;(rootMap.get(root)!.defineModel ??= []).push({
            expression: macroExpression,
            isRequired: expression.type === 'TSNonNullExpression',
          })
        } else if (options.defineStyle.alias.includes(macroName)) {
          const lang =
            macroExpression.callee.type === 'MemberExpression' &&
            macroExpression.callee.property.type === 'Identifier'
              ? macroExpression.callee.property.name
              : 'css'
          ;(rootMap.get(root)!.defineStyle ??= []).push({
            expression: macroExpression,
            isDeclaration: node.type === 'VariableDeclaration',
            lang,
          })
        } else if (options.defineSlots.alias.includes(macroName)) {
          rootMap.get(root)!.defineSlots = macroExpression
        } else if (options.defineExpose.alias.includes(macroName)) {
          rootMap.get(root)!.defineExpose = macroExpression
        }
      }
    },
    leave() {
      parents.shift()
    },
  })
  return rootMap
}

export function isFunctionalNode(node?: Node | null): node is FunctionalNode {
  return !!(
    node &&
    (node.type === 'ArrowFunctionExpression' ||
      node.type === 'FunctionDeclaration' ||
      node.type === 'FunctionExpression')
  )
}

export function getMacroExpression(
  node: Node,
  options: OptionsResolved,
): CallExpression | undefined {
  if (node.type === 'TSNonNullExpression') {
    node = node.expression
  }

  if (node.type === 'CallExpression') {
    if (
      node.callee.type === 'MemberExpression' &&
      node.callee.object.type === 'Identifier' &&
      node.callee.object.name === 'defineStyle'
    ) {
      return node
    } else if (
      node.callee.type === 'Identifier' &&
      [
        ...options.defineComponent.alias,
        ...options.defineSlots.alias,
        ...options.defineModel.alias,
        ...options.defineExpose.alias,
        ...options.defineStyle.alias,
      ].includes(node.callee.name!)
    ) {
      return node
    }
  }
}

export function getParamsStart(node: FunctionalNode, code: string): number {
  return node.params[0]
    ? node.params[0].start!
    : node.start! +
        (code.slice(node.start!, node.body.start!).match(/\(\s*\)/)?.index ||
          0) +
        1
}
