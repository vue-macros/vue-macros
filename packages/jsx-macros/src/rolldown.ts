import unplugin from '.'

export default unplugin.rolldown as typeof unplugin.rolldown
