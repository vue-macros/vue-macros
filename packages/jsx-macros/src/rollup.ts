import unplugin from '.'

export default unplugin.rollup as typeof unplugin.rollup
