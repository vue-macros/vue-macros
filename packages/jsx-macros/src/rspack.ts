import unplugin from '.'

export default unplugin.rspack as typeof unplugin.rspack
