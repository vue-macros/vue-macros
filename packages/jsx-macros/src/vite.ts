import unplugin from '.'

export default unplugin.vite as typeof unplugin.vite
