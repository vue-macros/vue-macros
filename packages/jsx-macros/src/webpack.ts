import unplugin from '.'

export default unplugin.webpack as typeof unplugin.webpack
