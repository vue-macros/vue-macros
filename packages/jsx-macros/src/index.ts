import { createUnplugin, type UnpluginInstance } from 'unplugin'
import { plugin, type Options, type OptionsResolved } from './core/plugin'

export type { Options, OptionsResolved }

const unplugin: UnpluginInstance<Options | undefined, false> =
  createUnplugin(plugin)
export default unplugin
