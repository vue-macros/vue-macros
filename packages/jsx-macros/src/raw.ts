import { plugin } from './core/plugin'

export default plugin
