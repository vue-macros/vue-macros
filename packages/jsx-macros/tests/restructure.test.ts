import { babelParse, MagicStringAST, walkAST } from '@vue-macros/common'
import { describe, expect, test } from 'vitest'
import { restructure } from '../src/api'

const transformRestructure = (code: string): string => {
  const s = new MagicStringAST(code)
  const ast = babelParse(code, 'tsx')
  walkAST(ast, {
    enter(node) {
      if (
        node.type === 'FunctionExpression' ||
        node.type === 'ArrowFunctionExpression' ||
        node.type === 'FunctionDeclaration'
      ) {
        restructure(s, node)
      }
    },
  })
  return s.toString()
}

describe('transform', () => {
  test('reconstruct', () => {
    const code = transformRestructure(
      `const App = ([[[,foo]], {id: {foo: [bar]}}], { baz }) => {
        function onClick({ foo }){
          return { foo, baz: baz.baz }
        };
        return [ foo, bar, baz ]
      }`,
    )!
    expect(code).toMatchSnapshot()
  })

  test('reconstruct arrowFunctionExpression', () => {
    const code = transformRestructure(
      `const App = ([{root: {foo}}]) => (
        <>{[foo]}</>
      )`,
    )!
    expect(code).toMatchSnapshot()
  })

  test('reconstruct default-prop', () => {
    const code = transformRestructure(
      `function App({foo: bar = 'bar', baz: qux, ...rest}, [foo = 'foo']){
        return <>{[bar, qux, rest, foo]}</>
      }`,
    )!
    expect(code).toMatchSnapshot()
  })

  test('reconstruct rest-prop', () => {
    const code = transformRestructure(
      `function App({foo, bar = 1, ...rest}){
        return <>{[foo, bar, rest]}</>
      }`,
    )!
    expect(code).toMatchSnapshot()
  })
})
