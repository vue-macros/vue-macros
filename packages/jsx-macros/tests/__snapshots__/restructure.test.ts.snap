// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transform > reconstruct 1`] = `
"const App = (__MACROS_props, __MACROS_props1) => {
        function onClick(__MACROS_props){
          return { foo: __MACROS_props.foo, baz: __MACROS_props1.baz.baz }
        };
        return [ __MACROS_props[0][0][1], __MACROS_props[1].id.foo[0], __MACROS_props1.baz ]
      }"
`;

exports[`transform > reconstruct arrowFunctionExpression 1`] = `
"const App = (__MACROS_props) => (
        <>{[__MACROS_props[0].root.foo]}</>
      )"
`;

exports[`transform > reconstruct default-prop 1`] = `
"
import { createPropsDefaultProxy as __MACROS_createPropsDefaultProxy } from "/vue-macros/jsx-macros/with-defaults";
import { createPropsRestProxy as __MACROS_createPropsRestProxy } from "vue";function App(__MACROS_props, __MACROS_props1){
const __MACROS_default_props = __MACROS_createPropsDefaultProxy(__MACROS_props, {'.foo': 'bar'});
const __MACROS_default_props1 = __MACROS_createPropsDefaultProxy(__MACROS_props1, {'[0]': 'foo'});
const rest = __MACROS_createPropsRestProxy(__MACROS_default_props, ['foo', 'baz']);
        return <>{[__MACROS_default_props.foo, __MACROS_default_props.baz, rest, __MACROS_default_props1[0]]}</>
      }"
`;

exports[`transform > reconstruct rest-prop 1`] = `
"
import { createPropsDefaultProxy as __MACROS_createPropsDefaultProxy } from "/vue-macros/jsx-macros/with-defaults";
import { createPropsRestProxy as __MACROS_createPropsRestProxy } from "vue";function App(__MACROS_props){
const __MACROS_default_props = __MACROS_createPropsDefaultProxy(__MACROS_props, {'.bar': 1});
const rest = __MACROS_createPropsRestProxy(__MACROS_default_props, ['foo', 'bar']);
        return <>{[__MACROS_default_props.foo, __MACROS_default_props.bar, rest]}</>
      }"
`;
